### PR TITLE
clarify `span_label` documentation

### LINKED ISSUE #71857: DiagnosticBuilder::span_label (sometimes?) does nothing
Recently, I was constructing a diagnostic with code like this:
```rust
            let mut diag = self.struct_warn("skipping const checks");
            for &(span, feature_gate) in unleashed_features.iter() {
                if let Some(feature_gate) = feature_gate {
                    diag.span_label(span, format!("skipping check for `{}` feature", feature_gate));
                } else {
                    diag.span_label(span, "skipping check that does not even have a feature gate");
                }
            }
            diag.emit();
```
I expected this to print a warning with a bunch of labels pointing to the given spans attached to it. But instead, the labels are just entirely ignored.

I am now using `span_help` instead, but that's a gross hack as this is not a "help" information. (This warning is for permanently unstable nightly-only testing flags, so it's not a big deal.) Unfortunately the `span_label` doc comment is also not very helpful, talking about the internal representation of the diagnostic builder ("multispan" etc) and about what happens when I do *not* call this method (?!?) but not about why adding a label might just have no effect at all.

Cc @rust-lang/wg-diagnostics which hopefully has people that actually understand this API, unlike me. :)

### PR DESCRIPTION
**Pull request by euclio**
Fixes #71857.

r? @estebank 

cc @RalfJung

### CODE REVIEW

File: compiler/rustc_errors/src/diagnostic_builder.rs [hunk 1]
```diff
@@ -184,16 +184,17 @@ impl<'a> DiagnosticBuilder<'a> {
...
+    /// be shown together with the original diagnostic's span, *not* with spans added by
+    /// `span_note`, `span_help`, etc. If the primary span is not displayable, labels will not be
+    /// displayed either.
     ///
-    /// This span is *not* considered a ["primary span"][`MultiSpan`]; only
-    /// the `Span` supplied when creating the diagnostic is primary.
-    ///
-    /// [`MultiSpan`]: ../rustc_span/struct.MultiSpan.html
+    /// Implementation-wise, the label span is pushed onto the [`MultiSpan`] that was created when
+    /// the diagnostic was constructed, though the label span is *not* considered a
```
Comment (RalfJung) [inline compiler/rustc_errors/src/diagnostic_builder.rs:195]:
```suggestion
    /// the diagnostic was constructed. However. the label span is *not* considered a
```

File: compiler/rustc_errors/src/diagnostic_builder.rs [hunk 2]
```diff
@@ -184,16 +184,17 @@ impl<'a> DiagnosticBuilder<'a> {
...
-    /// Adds a span/label to be included in the resulting snippet.
+    /// Appends a labeled span to the diagnostic.
     ///
-    /// This is pushed onto the [`MultiSpan`] that was created when the diagnostic
-    /// was first built. That means it will be shown together with the original
-    /// span/label, *not* a span added by one of the `span_{note,warn,help,suggestions}` methods.
+    /// Labels are used to convey additional context for the diagnostic's primary span. They will
+    /// be shown together with the original diagnostic's span, *not* with spans added by
+    /// `span_note`, `span_help`, etc. If the primary span is not displayable, labels will not be
+    /// displayed either.
```
Comment (RalfJung) [inline compiler/rustc_errors/src/diagnostic_builder.rs:192]: When is a span "not displayable"? Would be good to have an example.

Comment (estebank) [reply compiler/rustc_errors/src/diagnostic_builder.rs:192]: When they are `DUMMY_SP` or the source code for them isn't found (external crates or std on certain deployments where remap path isn't enabled).

**Comment by estebank** [comment]
r=me after addressing @RalfJung's comments.

**Comment by euclio** [comment]
Updated.

**Pull request merged**